# Analyze placementdata - 10k.csv
Clean dataset to have numbers and normalize each column

In [ ]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
datasetFileName = "..3/data/placementdata - 10k.csv"

placementDataFrame = pd.read_csv(datasetFileName)

placementDataFrame.head()


,StudentID,CGPA,Internships,Projects,Workshops/Certifications,AptitudeTestScore,SoftSkillsRating,ExtracurricularActivities,PlacementTraining,SSC_Marks,HSC_Marks,PlacementStatus
0,1,7.5,1,1,1,65,4.4,No,No,61,79,NotPlaced
1,2,8.9,0,3,2,90,4.0,Yes,Yes,78,82,Placed
2,3,7.3,1,2,2,82,4.8,Yes,No,79,80,NotPlaced
3,4,7.5,1,1,2,85,4.4,Yes,Yes,81,80,Placed
4,5,8.3,1,2,2,86,4.5,Yes,Yes,74,88,Placed


In [3]:
numericColumns = []
nonNumericColumns = []

for c in placementDataFrame.columns[1:]: #Exclude StudentID
    if(placementDataFrame[c].dtype == "object"):
        nonNumericColumns.append(c)
    else:
        numericColumns.append(c)

averages = {}
for c in numericColumns:
    l = len(placementDataFrame[c])
    s = 0
    for i in range(l):
        s += placementDataFrame[c][i]
    averages[c] = s/l

labelPercentages = {}
for c in nonNumericColumns:
    l = len(placementDataFrame[c])
    labels = {}
    for i in range(l):
        val = placementDataFrame[c][i]
        if val not in labels:
            labels[val] = 1
        else:
            labels[val] += 1
    
    s = 0
    for label in labels:
        s += labels[label]
    for label in labels:
        labels[label] = (labels[label]/s)*100

    labelPercentages[c] = labels


print("Averages:")
for c in averages:
    print(f"{c} - {averages[c]}")
print()
print("Percentages")


for c in labelPercentages:
    labels = labelPercentages[c]
    print(c)
    for label in labels:
        print(f"{label} - {labels[label]}", end="  ")
    print()

Averages:
CGPA - 7.69800999999997
Internships - 1.0492
Projects - 2.0266
Workshops/Certifications - 1.0132
AptitudeTestScore - 79.4499
SoftSkillsRating - 4.3239600000000395
SSC_Marks - 69.1594
HSC_Marks - 74.5015

Percentages
ExtracurricularActivities
No - 41.46  Yes - 58.540000000000006  
PlacementTraining
No - 26.82  Yes - 73.18  
PlacementStatus
NotPlaced - 58.03  Placed - 41.97  


# Data Analysis
## Averages:
- CGPA - 7.69800999999997
- Internships - 1.0492
- Projects - 2.0266
- Workshops/Certifications - 1.0132
- AptitudeTestScore - 79.4499
- SoftSkillsRating - 4.3239600000000395
- SSC_Marks - 69.1594
- HSC_Marks - 74.5015

## Percentages
### ExtracurricularActivities
 - No - 41.46  Yes - 58.540000000000006  
### PlacementTraining
 - No - 26.82  Yes - 73.18  
### PlacementStatus
 - NotPlaced - 58.03  Placed - 41.97 

In [4]:
# Mapping text to numbers

placementDataFrame["ExtracurricularActivities"] = (
    placementDataFrame["ExtracurricularActivities"]
    .str.strip()         
    .str.lower()         
    .map({"yes": 1, "no": 0})
)

placementDataFrame["PlacementTraining"] = (
    placementDataFrame["PlacementTraining"]
    .str.strip()         
    .str.lower()         
    .map({"yes": 1, "no": 0})
)

placementDataFrame["PlacementStatus"] = (
    placementDataFrame["PlacementStatus"]
    .str.strip()         
    .str.lower()         
    .map({"notplaced": 0 , "placed": 1})
)

# Normalizing all columns to be between 0 and 1
scaler = MinMaxScaler()
numeric_cols = ['CGPA', 'Internships', 'Projects', 'Workshops/Certifications',
                'AptitudeTestScore', 'SoftSkillsRating', 'SSC_Marks', 'HSC_Marks']

placementDataFrame[numeric_cols] = scaler.fit_transform(placementDataFrame[numeric_cols])


In [ ]:
# Drop StudentID (not a predictive feature)
X = placementDataFrame.drop(columns=["StudentID", "PlacementStatus", "AptitudeTestScore", "Workshops/Certifications", ])
y = placementDataFrame["PlacementStatus"]

print(X.head())

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train model
model = LogisticRegression()
model.fit(X_train, y_train)

# Predict and evaluate
y_pred = model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


       CGPA  Internships  Projects  SoftSkillsRating  \
0  0.384615          0.5  0.333333          0.777778   
1  0.923077          0.0  1.000000          0.555556   
2  0.307692          0.5  0.666667          1.000000   
3  0.384615          0.5  0.333333          0.777778   
4  0.692308          0.5  0.666667          0.833333   

   ExtracurricularActivities  PlacementTraining  SSC_Marks  HSC_Marks  
0                          0                  0   0.171429   0.709677  
1                          1                  1   0.657143   0.806452  
2                          1                  0   0.685714   0.741935  
3                          1                  1   0.742857   0.741935  
4                          1                  1   0.542857   1.000000  
Accuracy: 0.785
              precision    recall  f1-score   support

           0       0.83      0.80      0.81      1172
           1       0.73      0.76      0.75       828

    accuracy                           0.79      20

## Different Models Comparison for Placement Data

### Logistic Regression Model
- 79% accuracy with Logistic Regression Model excluding features which are dropped

---

# Analyse student career suggestion - 20k.csv

In [ ]:
import pandas as pd
# from collections import Counter

suggestionDataFileName = "../data/student career suggestion - 20k.csv"

suggestionDF = pd.read_csv(suggestionDataFileName)

# role_count = Counter(suggestionDF["Suggested Job Role"])

# print(len(role_count))


34


In [2]:
# Change column names
suggestionDF.columns = ['Operating Systems', 'Algorithms',
       'Programming Concepts',
       'Software Engineering', 'Computer Networks',
       'Electronics Subjects',
       'Computer Architecture', 'Mathematics',
       'Communication Skills', 'Hours working per day',
       'Logical quotient rating', 'hackathons', 'coding skills rating',
       'public speaking points', 'can work long time before system?',
       'self-learning capability?', 'Extra-courses did', 'certifications',
       'workshops', 'talenttests taken?', 'olympiads',
       'reading and writing skills', 'memory capability score',
       'Interested subjects', 'interested career area ', 'Job/Higher Studies?',
       'Type of company want to settle in?',
       'Taken inputs from seniors or elders', 'interested in games',
       'Interested Type of Books', 'Salary Range Expected',
       'In a Realtionship?', 'Gentle or Tuff behaviour?',
       'Management or Technical', 'Salary/work', 'hard/smart worker',
       'worked in teams ever?', 'Introvert', 'Suggested Job Role']



# Dataset Stats (student career suggestion - 20k.csv)

34 total unique job roles
- Database Developer -> 581
- Portal Administrator -> 593
- Systems Security Administrator -> 562
- Business Systems Analyst -> 582
- Software Systems Engineer -> 575
- Business Intelligence Analyst -> 540
- CRM Technical Developer -> 567
- Mobile Applications Developer -> 538
- UX Designer -> 589
- Quality Assurance Associate -> 565
- Web Developer -> 570
- Information Security Analyst -> 543
- CRM Business Analyst -> 584
- Technical Support -> 565
- Project Manager -> 602
- Information Technology Manager -> 591
- Programmer Analyst -> 529
- Design & UX -> 588
- Solutions Architect -> 578
- Systems Analyst -> 550
- Network Security Administrator -> 1112
- Data Architect -> 564
- Software Developer -> 587
- E-Commerce Analyst -> 546
- Technical Services/Help Desk/Tech Support -> 558
- Information Technology Auditor -> 558
- Database Manager -> 570
- Applications Developer -> 551
- Database Administrator -> 593
- Network Engineer -> 621
- Software Engineer -> 590
- Technical Engineer -> 557
- Network Security Engineer -> 630
- Software Quality Assurance (QA) / Testing -> 571


In [9]:
# Test Random Forest

X = suggestionDF.drop('Suggested Job Role', axis=1)  # features
y = suggestionDF['Suggested Job Role']              # target

X = pd.get_dummies(X, dtype=int)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)

from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)

from sklearn.metrics import accuracy_score

y_pred = rf.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred)*100)


Accuracy: 3.4166666666666665


# Different Models for Job Role Suggestion
34 classifications with 126 features
- Random Forest - 3.416% Accuracy